### ISSUE TITLE
Removable Storage async File Write Parity

### ISSUE THREAD

**Issue by NjlsShade** [elementary/files#1748]
## Problem
When removable storage devices are mounted with the default async option Linux file managers generally show the progress of transfer operations to said device as concluding once all data has been copied to the memory buffer, this is great for internal drives, as it frees up the files app to do other things while the operation continues in the background. However, the nature of removable media renders this just not a practical choice.

On slower storage devices this can cause confusion for inexperienced users and is a general inconvenience for anyone, as even if you unmount the device via the Files app while data is still being written from the buffer, there is no clear indication the system is doing so, nor is there a vivid confirmation on the completion of said data write (other then the drive disappearing from the Storage list, however, this is subtle, inconsistent, and can be missed).

## Proposal
File transfers to removable media should follow the progress of the full async task via the progress bar, not just the copy to buffer. Optionally provide a button, check, or setting to force a sync operation if the user desires, making sure to state that this will write data the fastest speed possible at the expense of the device's longevity.

Also, providing clear notifications when a device has been unmounted successfully or removed incorrectly (without prior umounting) would go a long way as well.

## Prior Art
From what I gather, Chromium OS visulaises the progress of removable media's async transfer via the file copy progress bar, thus eliminating this consistency issue. I do not know for certain, however, I would assume both Mac OS and Windows do the same as their file transfers also keep parity with the actual data being written to external hardware.

As someone who conducts file transfers regularly It's tedious to have to manually monitor the block device's dirty bytes for progress every time I write a large sum of data. Would love this feature to be included in some form, really enjoying Elementary OS 6 Beta so far and look forward to using it as my daily driver once released. Thanks!

**Comment by jeremypw** [comment]
I agree this is a problem but as far as I am aware, the copy to memory buffer happens transparently in the underlying gvfs system. It is not a Files function. I am not sure there is a way of Files knowing whether the data has actually been written to drive or has just been cached.

**Comment by NjlsShade** [comment]
After looking for some time I can't seem to find a way to monitor the progress of dirty bytes being written to disk "per device", however, unless I'm wrong, mounting a drive with the sync option only negatively affects older FAT filesystems, where exFAT, NTFS and Linux File Systems all perform no extra work writing the data synchronously.

If that is the case we could have all removable media mounted with the sync option by default, while all older FAT filesystems simply mount using the flush option. This would fix the issue for the majority of filesystems, only downside would be that copy dialogs for FAT drives would hang at like ~90% copied until the data has been written completely (which is arguably better than forcing sync to run be ejecting it?).

Not something Files is in charge of though I assume.

**Comment by vjr** [comment]
@jeremypw (or perhaps anyone @elementary/desktop-developers) can we revisit this open issue for Files for os8?

For the scenario of an "everyday" user (hello non-tech-savvy grandparents out there!) who might plug in a usb pen drive to copy some file(s) they observe (like me) the progress bar window completes and closes and then they immediately unplug the drive because unbeknownst to them they are supposed to click the eject/unmount button.

How about always flushing the destination file handle, such as adding the following (chatgpt suggested) example code snippet into this location: https://github.com/elementary/files/blob/main/libcore/marlin-file-operations.c#L2282 ?

```
// Open the destination file for flushing
output_stream = G_OUTPUT_STREAM(g_file_replace(
    dest, 
    NULL, 
    FALSE, 
    G_FILE_CREATE_NONE, 
    NULL, 
    NULL
));

if (output_stream) {
    // Flush the output stream to ensure data is written
    g_output_stream_flush(output_stream, NULL, NULL)
}
```

I haven't gotten around to trying this myself to see how it behaves with existing copies for internal/fixed drives and removable drives but wanted to get any input on whether this is worth pursuing.

**Comment by jeremypw** [comment]
Some work has been done converting the file operations to Vala (thanks to @tintou) and this should be considered in that process.
